File: src/backend/models/purchaseRequestModel.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const purchaseRequestSchema = new Schema({
    advertisement: {
        type: Schema.Types.ObjectId,
        ref: 'Advertisement',
        required: true
    },
    buyer: {
        type: Schema.Types.ObjectId,
        ref: 'Buyer',
        required: true
    },
    quantity: {
        type: Number,
        required: true,
        min: [0.1, 'Quantity must be greater than 0']
    },
    unit: {
        type: String,
        required: true,
        enum: ['kg', 'g', 'L', 'ml', 'unit', 'units']
    },
    status: {
        type: String,
        enum: ['pending', 'accepted', 'rejected', 'cancelled', 'paid'],
        default: 'pending'
    },
    rejectionStatus: {
        type: String,
        maxlength: [50, 'Description cannot exceed 50 characters']
    },
    pricePerUnit: {
        type: Number
    },
    totalPrice: {
        type: Number
    },
    wantToImportThem :{
        type: Boolean,
        required: true,
        default: false
    },
    deliveryLocation: String,
    createdAt: {
        type: Date,
        default: Date.now
    }
}, { timestamps: true });

// Auto-calculate price before saving
purchaseRequestSchema.pre('save', async function (next) {
  const Advertisement = mongoose.model('Advertisement');
  const ad = await Advertisement.findById(this.advertisement);
  if (!ad) return next(new Error('Invalid advertisement'));

  const matchedTier = ad.priceTiers.find(tier =>
    this.quantity >= tier.minQuantity && this.quantity <= tier.maxQuantity
  );

  if (!matchedTier) {
    return next(new Error('Requested quantity doesn’t match any price tier'));
  }

  this.pricePerUnit = matchedTier.price;
  this.totalPrice = this.quantity * matchedTier.price;

  next();
});


const PurchaseRequest = mongoose.model('PurchaseRequest', purchaseRequestSchema);
module.exports = PurchaseRequest;

File: src/frontend/lib/models/purchase_request_model.dart
import 'dart:convert';

import 'package:frontend/models/tender_model.dart';

enum Status { pending, accepted, rejected, cancelled, paid }

String statusToString(Status status) {
  switch (status) {
    case Status.pending:
      return 'pending';
    case Status.accepted:
      return 'accepted';
    case Status.rejected:
      return 'rejected';
    case Status.cancelled:
      return 'cancelled';
    case Status.paid:
      return 'paid';
  }
}

Status statusFromString(String status) {
  switch (status) {
    case 'pending':
      return Status.pending;
    case 'accepted':
      return Status.accepted;
    case 'rejected':
      return Status.rejected;
    case 'cancelled':
      return Status.cancelled;
    case 'paid':
      return Status.paid;
    default:
      throw Exception('Invalid Status');
  }
}

class PurchaseRequestModel {
  String? id;
  dynamic advertisement;
  dynamic buyerId;
  final double quantity;
  Units? unit;
  Status? status;
  String? rejectionStatus;
  double? pricePerUnit;
  double? totalPrice;
  bool? wantToImportThem;
  String? deliveryLocation;
  DateTime? createdAt;

  PurchaseRequestModel({
    this.id,
    this.advertisement,
    this.buyerId,
    required this.quantity,
    this.unit,
    this.status,
    this.rejectionStatus,
    this.pricePerUnit,
    this.totalPrice,
    this.wantToImportThem,
    this.deliveryLocation,
    this.createdAt,
  });

  // Helper method to get Advertisement Name
  String get advertisementName {
    if (advertisement is String) return advertisement;
    if (advertisement is Map<String, dynamic>) return advertisement['name'];
    return '';
  }

  // Helper method to get Buyer Name
  String get buyerName {
    if (buyerId is String) return buyerId;
    if (buyerId is Map<String, dynamic>) {
      return buyerId['firstName'] + buyerId['lastName'];
    }
    return '';
  }

  factory PurchaseRequestModel.fromJson(Map<String, dynamic> json) {
    return PurchaseRequestModel(
      id: json['_id'],
      advertisement: json['advertisement'],
      buyerId: json['buyer'],
      quantity: json['quantity'].toDouble(),
      unit: unitsFromString(json['unit']),
      status: statusFromString(json['status']),
      rejectionStatus: json['rejectionStatus'],
      pricePerUnit: json['pricePerUnit'].toDouble(),
      totalPrice: json['totalPrice'].toDouble(),
      wantToImportThem: json['wantToImportThem'],
      deliveryLocation: json['deliveryLocation'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {'quantity': quantity, 'rejectionStatus': rejectionStatus};
  }

  String toJsonString() => json.encode(toJson());
}
